web-server/web-server.c

```diff
@@ -0,0 +1,382 @@
+#include "web-server.h"
+
+#include <espconn.h>
+
+#include "espfs.h"
+#include "config.h"
+#include "cgi.h"
+#include "cmd.h"
+#include "serbridge.h"
+
+#define WEB_CB "webCb"
+
+#define MAX_VARS 20
+
+static char* web_server_reasons[] = {
+  "load", "refresh", "button", "submit"
+};
+
+char * webServerPages = NULL;
+
+char * ICACHE_FLASH_ATTR WEB_UserPages()
+{
+	return webServerPages;
+}
+
+void ICACHE_FLASH_ATTR WEB_BrowseFiles()
+{
+	char buffer[1024];
+	buffer[0] = 0;
+	
+	if( espFsIsValid( userPageCtx ) )
+	{
+		EspFsIterator it;
+		espFsIteratorInit(userPageCtx, &it);
+		while( espFsIteratorNext(&it) )
+		{
+			int nlen = os_strlen(it.name);
+			if( nlen >= 6 )
+			{
+				if( os_strcmp( it.name + nlen-5, ".html" ) == 0 )
+				{
+					char sh_name[17];
+					
+					int spos = nlen-5;
+					
+					while( spos > 0 )
+					{
+						if( it.name[spos+1] == '/' )
+							break;
+						spos--;
+					}
+					
+					int ps = nlen-5-spos;
+					if( ps > 16 )
+						ps = 16;
+					os_memcpy(sh_name, it.name + spos, ps);
+					sh_name[ps] = 0;
+					
+					os_strcat(buffer, ", \"");
+					os_strcat(buffer, sh_name);
+					os_strcat(buffer, "\", \"/");
+					os_strcat(buffer, it.name);
+					os_strcat(buffer, "\"");
+				}
+			}
+			if( os_strlen(buffer) > 600 )
+				break;
+		}
+	}
+	
+	if( webServerPages != NULL )
+		os_free( webServerPages );
+	
+	int len = os_strlen(buffer) + 1;
+	webServerPages = (char *)os_malloc( len );
+	os_memcpy( webServerPages, buffer, len );
+}
+
+void ICACHE_FLASH_ATTR WEB_Init()
+{
+	espFsInit(userPageCtx, (void *)getUserPageSectionStart(), ESPFS_FLASH);
+	if( espFsIsValid( userPageCtx ) )
+		os_printf("Valid user file system found!\n");
+	else
+		os_printf("No user file system found!\n");
+	WEB_BrowseFiles();
+}
+
+int ICACHE_FLASH_ATTR WEB_CgiJsonHook(HttpdConnData *connData)
+{
+	if (connData->conn==NULL) return HTTPD_CGI_DONE; // Connection aborted. Clean up.
+	
+	void * cgiData = connData->cgiData;
+	
+	if( cgiData == NULL )
+	{
+		if( !flashConfig.slip_enable )
+		{
+			errorResponse(connData, 400, "Slip processing is disabled!");
+			return HTTPD_CGI_DONE;
+		}
+		CmdCallback* cb = cmdGetCbByName( WEB_CB );
+		if( cb == NULL )
+		{
+			errorResponse(connData, 500, "No MCU callback is registered!");
+			return HTTPD_CGI_DONE;
+		}
+		if( serbridgeInProgramming() )
+		{
+			errorResponse(connData, 500, "Slip disabled at programming mode!");
+			return HTTPD_CGI_DONE;
+		}
+		
+		char reasonBuf[16];
+		int i;
+		int len = httpdFindArg(connData->getArgs, "reason", reasonBuf, sizeof(reasonBuf));
+		if( len < 0 )
+		{
+			errorResponse(connData, 400, "No reason specified!");
+			return HTTPD_CGI_DONE;
+		}
+		
+		RequestReason reason = INVALID;
+		for(i=0; i < sizeof(web_server_reasons)/sizeof(char *); i++)
+		{
+			if( os_strcmp( web_server_reasons[i], reasonBuf ) == 0 )
+				reason = (RequestReason)i;
+		}
+		
+		if( reason == INVALID )
```

Comment:
SLIP protocol contains a reason why esp-link initiated the SLIP query.

- LOAD -> when user loads a page at the first time
- REFRESH -> called after form submitting or periodical page reloading
- BUTTON -> when user presses an HTML button (containing the id of the button)
- SUBMIT -> when user submits a complete form (containing each data in the form)

As for refresh and load:
- imagine a form, that has a text field
- the normal behaviour is, that the content of the field is filled at loading the page
- user starts editing the field
- a page refresh happens (submitting another form on the same page) 
- it's not desired to restore thr field, that user started editing

So LOAD is for loading first time, REFRESH is for refreshing. There are controls where automated refreshing is not desired.